### PR TITLE
Mix release closure size

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

@dlesl started the work of removing the reference to erlang in rebarRelx resulting closure in https://github.com/NixOS/nixpkgs/pull/124779

This PR aims at starting the work for mix-release. @NobbZ you noticed in the initial PR that the resulting closure built by mix-release contained reference to erlang.
This PR does the following.
- Removing the setup hook from the generic builder of erlang. It seems it has no effect. Also from a reference found by @dlesl , it seems not needed https://erlang.org/doc/man/code.html
- Switches some dependencies from buildInputs to nativeBuildInputs. This is more a semantic difference, but just to make the intent clearer to future contributors.
- Remove reference to rebar into mix release. There is likely no need anymore for the old rebar.
- Remove the explicit reference of erlang into closure built by mix-release.

Please note however that the resulting closure built by mix-release still reference erlang somehow, even though I can't find any explicit reference by grepping. This is still a work in progress.

Since removing the erlang setup hook might be impactful, I'm curious to have people test this and give feedback

@NixOS/beam 

@ydlr @petabyteboy @cw789 you might be interested in this as well.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/beam-modules/mix-release.nix [hunk 1]
```diff
@@ -23,15 +23,14 @@ let
 in
 assert mixNixDeps != { } -> mixFodDeps == null;
 stdenv.mkDerivation (overridable // {
-  nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir makeWrapper git ];
+  nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir makeWrapper git gnused ];
```
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/mix-release.nix:26]: Isn't sed in the default stdenv?

Comment (happysalada) [reply pkgs/development/beam-modules/mix-release.nix:26]: good catch, I forgot to remove it

**Comment by happysalada** [comment]
TIL `nix why-depends`
I was browsing randomly through the doc and discovered this.
Needless to say it works!
```
❯ nix why-depends ./result ../../nixpkgs#erlang
/nix/store/l0339q9yqnqya6al8ay0b3bmg9m3rmvv-union-0.0.1
└───lib/bcrypt_elixir-2.3.0/priv/bcrypt_nif.so.dSYM/Contents/Resources/DWARF/bcrypt_nif.so: ….............
    → /nix/store/k5vq393x6c0ykfwamwcy0jhlf6iw9d8w-erlang-24.0.2
```

**Comment by dlesl** [comment]
Nice find! It might just be debug info. I think we might need to modify `buildRebar3` and `buildMix` to strip ELF/DWARF binaries in `/priv`, because I don't think the default builder looks there, see [`stripDebugList`](https://nixos.org/manual/nixpkgs/stable/#ssec-fixup-phase). It could also be a linux/darwin difference, I'll test some NIF-containing derivations on linux.

**Comment by dlesl** [comment]
I was able to reproduce this with `buildRebar3`. Since the output goes to `$out/lib` which is already in `stripDebugList`, all I had to do was remove the line `dontStrip = true`

**Comment by dlesl** [comment]
Would you be able to make the `dontStrip` change to `buildRebar3` as part of this PR (since packages built by it might end up as dependencies of `mixRelease` derivations), and I can make a separate PR to fix `rebar3Relx` (since its output goes to `rel`, I think we will need to override `stripDebugList`)?

**Comment by happysalada** [comment]
Thanks for the find on stripping.
From what I understood stripping has no effect on beam files. That's why originally `dontStrip` was set to true.
If I understand well, for buildMix it will be useful for dependencies with NIFs so I added `stripDebugList = [ "priv" ];`
with the example of bcrypt that has stuff to be stripped in priv
No problem in removing the `dontStrip` for buildRebar3 in this PR.

**Comment by dlesl** [comment]
Thanks! Since `buildMix` also copies everything into `$out/lib/erlang/lib` you probably don't need to set `stripDebugList`?

**Comment by happysalada** [comment]
@dlesl added the commit that references stripping. Let me know if you think it's a little too much to add a comment and leave the default.

Unless I'm wrong, even after being stripped, the reference to erlang remains. So I still need to do some grep magic to remove it in the postFix hook of mix-release. Since I think you only know at that time if ERTS is present or not. So probably not possible to do it in a buildMix hook.

**Comment by happysalada** [comment]
Also just to make sure, this is how I tested that the bcrypt.so binary was stripped
`objdump --syms bcrypt_nif.so | grep debug`
just in case it wasn't the right way.

**Comment by dlesl** [comment]
I think we should try to strip the debug info whether or not ERTS is included in the release, since stripping binaries is the default nix behaviour and leaving it might cause other issues.

Regarding the reference to erlang remaining, it might be a difference between darwin and linux? IMHO it's good enough if it works correctly on linux (just guessing that most people deploy their releases on linux)

**Comment by happysalada** [comment]
> I think we should try to strip the debug info whether or not ERTS is included in the release, since stripping binaries is the default nix behaviour and leaving it might cause other issues.

I agree, did I mistakenly include some conditional in there? Or are you talking about stripping FOD deps?
including the stripping in buildMix and buildRebar3 should be enough right? Let me know if I missed something.

> Regarding the reference to erlang remaining, it might be a difference between darwin and linux? IMHO it's good enough if it works correctly on linux (just guessing that most people deploy their releases on linux)

I don't think it's a darwin specific thing unfortunately. I've pushed a fix. 
Here is the result
After
```
❯ nix path-info -rsSh ./result
/nix/store/4salp61q5ic2zw6s22ii0j8d7dxg1vfx-union-0.0.1                   38.9M  117.9M
/nix/store/4salp61q5ic2zw6s22ii0j8d7dxg1vfx-union-0.0.1                   38.9M  117.9M
/nix/store/64d69jqbz4s8ziqbpam41sd70w338ars-libcxx-7.1.0                   2.4M    2.9M
/nix/store/6q1rhqvyzxb3j5pb1h4qxsa0hh1i45wv-libcxxabi-7.1.0              533.3K  533.3K
/nix/store/892bqs4h7gck4qzysnj6s8niy9jk9qbn-libkrb5-1.18                   1.6M   19.8M
/nix/store/8f4wp5sdl08wrnj3ajsqrjbv7fw5lgv7-zlib-1.2.11                  102.2K  102.2K
/nix/store/96j2ai9h4f5kaqcnp9kvgbp2srr79xhx-openssl-1.1.1k                 3.1M   75.9M
/nix/store/az26q1v080bqfwhn8snapbscxgdw80qn-nghttp2-1.43.0-lib           167.4K  167.4K
/nix/store/cykikdmpsksd35xs8ppds7iwqa2rllzl-curl-7.76.1                  560.8K   25.8M
/nix/store/d4spin3lyclblj5lji2ih4axg19dfmfw-libssh2-1.9.0                233.5K    3.4M
/nix/store/f695vk4n50d3i528ldnf9rvnkciknm5y-libxml2-2.9.12                 1.4M    1.5M
/nix/store/f7f3gz636cqxkav7cy1nlldkwv2b22n1-swift-corefoundation           3.3M   72.8M
/nix/store/kdydav4gqzd91qlpprnjascbs5hs8qmg-ncurses-6.2                    3.2M    3.2M
/nix/store/npjd5jnk4chfz39lv0vm24f029ry6h6r-openssl-1.1.1k                 3.1M    3.1M
/nix/store/pcjan45rssdn01cxx3sjg70avjg6c3ni-bash-4.4-p23                 997.6K  997.6K
/nix/store/qhz9xpqbj6xvwmzxwspdixpr4px2l22j-ICU-66108                     39.4M   42.2M
/nix/store/riasnjlb9bb3d3gciv2aaz1ypxrg7hp3-brotli-1.0.9-lib               1.9M    1.9M
/nix/store/rvqdp9l0ncp69mcxfpf0vfmrafh8zqa8-Libsystem-1238.60.2           17.2M   17.2M
```

**Comment by happysalada** [comment]
@dlesl you were right about stripping things from buildMix, no need to override the stripDebug flag

It made me realize that the mix-release also needs to be stripped. I think that is what you were referring to earlier.
I'm testing this at the moment.

**Comment by happysalada** [comment]
I changed the fixupPhase to a postFixup since there is no use overriding anymore.
I think this PR is good to merge.
Will leave one more day open in case anybody wants to comment or test.

**Pull request merged**